File: Sharebook/Server/src/Services/Upload/Sharebook.Upload.API/Wrappers/JObjectStaticWrapper.cs
namespace Sharebook.Upload.API.Wrappers
{
    using Newtonsoft.Json;
    using Newtonsoft.Json.Linq;

    using Sharebook.Common.Extensions;
    using Sharebook.Common.Wrappers;
    using Sharebook.Upload.API.JsonConverters;

    // Wrapper classes "cannot" be tested
    public class JObjectStaticWrapper : IJObjectStaticWrapper
    {
        private readonly JsonSerializer serializer;

        public JObjectStaticWrapper()
        {
            this.serializer = new JsonSerializer();

            foreach (var converter in JsonPropertiesConverterRepository.Converters)
            {
                this.serializer.Converters.Add(converter);
            }
        }

        public JObject GetJobjectFromJson(string json)
        {
            if (!StringExtensions.IsMatchingJsonFormat(json)) // TODO: might bug, as it was !StringExtension
            {
                return null;
            }

            return JObject.Parse(json);
        }

        public T ToObject<T>(JObject jObject)
        {
            return jObject.ToObject<T>(this.serializer);
        }

        public T ToObject<T>(JToken jToken) where T : class
        {
            if (jToken == null)
            {
                return null;
            }

            return jToken.ToObject<T>(this.serializer);
        }

        public string GetNodeRootName(JToken jToken)
        {
            var jProperty = jToken as JProperty;
            return jProperty != null ? jProperty.Name : null;
        }
    }
}

File: Sharebook/Client/src/app/logic/Wrapper/JObjectStaticWrapper.ts

    // Wrapper classes "cannot" be tested
    export class JObjectStaticWrapper implements IJObjectStaticWrapper
    {
        private readonly _serializer: JsonSerializer;

        constructor()
        {
            this._serializer = new JsonSerializer();

            for (let converter of JsonPropertiesConverterRepository.Converters) {
                this._serializer.Converters.Add(converter);
            }
        }

        public  GetJobjectFromJson(json: string): JObject
        {
            if (!StringExtension.IsMatchingJsonFormat(json))
            {
                return null;
            }

            return JObject.Parse(json);
        }

        public  ToObject<T>(jObject: JObject): T
        {
            return jObject.ToObject<T>(this._serializer);
        }

        public  ToObject<T>(jToken: JToken): T
        {
            if (jToken == null)
            {
                return null;
            }

            return jToken.ToObject<T>(this._serializer);
        }

        public  GetNodeRootName(jToken: JToken): string
        {
            var jProperty = jToken as JProperty;
            return jProperty != null ? jProperty.Name : null;
        }
    }
